src/uffs/uffs_tree.c

```diff
@@ -1082,6 +1085,9 @@ static URET _BuildTreeStepThree(uffs_Device *dev)
 	uffs_Perror(UFFS_MSG_NOISY, "build tree step three");
 
 	for (i = 0; i < DATA_NODE_ENTRY_LEN; i++) {
+              if (dev->OnBuildTreeBlockProcessed != NULL) {
```

Comment:
You are executing the callback during step one and during step three. This means callback may be called for some (all?) blocks twice.

It may be better to provide separate callbacks for both steps.

Also, consider adding a block number as a callback argument.

Reply:
Added second callback and added block number as argument for both.